wandb/integration/xgboost/xgboost.py

```diff
@@ -0,0 +1,189 @@
+    """Old style callback that will be deprecated in favor of WandbCallback. Please try the new logger for more features."""
+    warnings.warn(
+        "wandb_callback will be deprecated in favor of WandbCallback. Please use WandbCallback for more features.",
+        UserWarning,
+        stacklevel=2,
+    )
+
+    with wb_telemetry.context() as tel:
+        tel.feature.xgboost_old_wandb_callback = True
+
+    def callback(env: "CallbackEnv") -> None:
+        for k, v in env.evaluation_result_list:
+            wandb.log({k: v}, commit=False)
+        wandb.log({})
+
+    return callback
+
+
+class WandbCallback(xgb.callback.TrainingCallback):
+    """`WandbCallback` automatically integrates XGBoost with wandb.
+
+    Arguments:
+        log_model: (boolean) if True save and upload the model to Weights & Biases Artifacts
+        log_feature_importance: (boolean) if True log a feature importance bar plot
+        importance_type: (str) one of {weight, gain, cover, total_gain, total_cover} for tree model. weight for linear model.
+        define_metric: (boolean) if True (default) capture model performance at the best step, instead of the last step, of training in your `wandb.summary`.
+
+    Passing `WandbCallback` to XGBoost will:
+
+    - log the booster model configuration to Weights & Biases
+    - log evaluation metrics collected by XGBoost, such as rmse, accuracy etc to Weights & Biases
+    - log training metric collected by XGBoost (if you provide training data to eval_set)
+    - log the best score and the best iteration
+    - save and upload your trained model to to Weights & Biases Artifacts (when `log_model = True`)
+    - log feature importance plot when `log_feature_importance=True` (default).
+    - Capture the best eval metric in `wandb.summary` when `define_metric=True` (default).
+
+    Example:
+        ```python
+        bst_params = dict(
+            objective ='reg:squarederror',
+            colsample_bytree = 0.3,
+            learning_rate = 0.1,
+            max_depth = 5,
+            alpha = 10,
+            n_estimators = 10,
+            tree_method = 'hist'
+        )
+
+        xg_reg = xgb.XGBRegressor(**bst_params)
+        xg_reg.fit(X_train,
+            y_train,
+            eval_set=[(X_test, y_test)],
+            callbacks=[WandbCallback()])
+        )
+        ```
+    """
+
+    def __init__(
+        self,
+        log_model: bool = False,
+        log_feature_importance: bool = True,
+        importance_type: str = "gain",
+        define_metric: bool = True,
+    ):
+
+        self.log_model: bool = log_model
+        self.log_feature_importance: bool = log_feature_importance
+        self.importance_type: str = importance_type
+        self.define_metric: bool = define_metric
+
+        with wb_telemetry.context() as tel:
+            tel.feature.xgboost_wandb_callback = True
+
+        if wandb.run is None:
+            raise wandb.Error("You must call wandb.init() before WandbCallback()")
```

Comment:
```suggestion
        if wandb.run is None:
            raise wandb.Error("You must call wandb.init() before WandbCallback()")

        with wb_telemetry.context() as tel:
            tel.feature.xgboost_wandb_callback = True
```

cant add telemetry without the run. want to reorder these.